app/main.py

```diff
@@ -1,214 +1,51 @@
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
-
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
-
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
-
-    # mordred
-    mordred = knightsConfig["mordred"]
-
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
-
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
+from app.knight import Knight
+
+
+def health_point_check(knight: dict) -> None:
+    if knight["hp"] < 0:
+        knight["hp"] = 0
+
+
+def battle(knights_config: dict) -> dict:
+
+    hero_dict = {}
+    for person in knights_config.values():
+        hero = person
+        knight = Knight(
+            name=hero["name"],
+            power=hero["power"],
+            hp=hero["hp"]
+        )
+        knight.armour_defence(hero["armour"])
+        knight.use_of_weapons(hero["weapon"])
+        knight.apply_potion(hero["potion"])
+        hero["hp"] = knight.hp
+        hero["power"] = knight.power
+        hero["protection"] = knight.protection
+        new_dict = {hero["name"]: hero}
+        hero_dict.update(new_dict)
+
+    lancelot = hero_dict["Lancelot"]
+    arthur = hero_dict["Artur"]
+    mordred = hero_dict["Mordred"]
+    red_knight = hero_dict["Red Knight"]
 
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
-
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
-
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
     lancelot["hp"] -= mordred["power"] - lancelot["protection"]
```

Comment:
1. Create a function for this. 
2. It does not the correct solution. Use OOP not dicts. 
Why do you create objects and again convert to dicts?